### PR TITLE
Add extended kernel generator docs

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Adds some documentation on kernel generator, with emphasis on how to implement a new operation.

## Tests
No new tests. This is just documentation.

## Side Effects
None.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by SteveBronder** [comment]
For doc stuff I thought it might be easier just to make a gist with changes. How do you feel about the following. Mostly it makes the requirements into a listicle and rewords the first paragraph. The list makes the output doxygen a bit prettier

https://gist.github.com/SteveBronder/0c94b465dbf4ef9dbccd3fedddae4d6d

**Comment by t4c1** [comment]
Content is fine. I just don't understand why do you think GIST is preferable to a PR?

**Comment by SteveBronder** [comment]
Actually maybe that was silly.

One other thing about this is that it feels like these docs are more for how to add new expressions than ops right? For instance adding an unary or binary op is much simpler than what's in the docs here

**Comment by t4c1** [comment]
Right. Adding a new binary/unary op in fact much simpler, so I feel they don't need in depth explanation. What needs to be done can be easily deduced from code for existing ops.

File: stan/math/opencl/kernel_generator.hpp [hunk 1]
```diff
@@ -2,6 +2,107 @@
...
+ * an object derived from `operation_cl`. Such an object holds arguments of the
+ * operations as well as meta information needed to generate calculations on the
+ * arguments. Arguments to operations can be other operations, scalars
+ * or `matrix_cl` objects. An operation is evaluated when either an operation is
+ * assigned to a `matrix_cl` or a left-hand-side operation or `.eval()` is
+ * called.
+ *
+ * ## Defining a new kernel generator operation
+ *
+ * New kernel generator classes must satsify the conditions below:
```
Comment (SteveBronder) [inline stan/math/opencl/kernel_generator.hpp]:
```suggestion
 * ## Defining a new kernel generator expression
 *
 * For adding new Unary and Binary ops use the macros provided in
 * `unary_function_cl.hpp` and `binary_operation.hpp`.
 *
 * New kernel generator expressions classes must satisfy the conditions below:
```

Comment (t4c1) [reply stan/math/opencl/kernel_generator.hpp]: Macro for binary operation only works for OpenCL operators. I think I covered all of them.

**Pull request merged**